### PR TITLE
Revert "Revert "Add mingwW64-llvm cross-system.""

### PR DESCRIPTION
**Pull request by shlevy**
Waiting on next staging cycle to move forward.

This reverts commit c911240e9cc2a7c12c985bd32700b9657b5698a9.

### CODE REVIEW

**Comment by shlevy** [comment]
@sternenseemann Any sense of when you'll get a chance to review?

File: pkgs/development/compilers/llvm/13/bintools/default.nix [hunk 1]
```diff
@@ -1,11 +1,17 @@
...
   prefix =
     if stdenv.hostPlatform != stdenv.targetPlatform
     then "${stdenv.targetPlatform.config}-"
     else "";
-in runCommand "llvm-binutils-${version}" { preferLocalBuild = true; } ''
+in runCommand "llvm-binutils-${version}" {
+  preferLocalBuild = true;
+  passthru = {
+    isLld = true;
+    targetPrefix = prefix;
```
Comment (sternenseemann) [inline pkgs/development/compilers/llvm/13/bintools/default.nix:12]: Why don't the other LLVM versions have a `targetPrefix` passthru now?

File: pkgs/development/compilers/llvm/13/bintools/default.nix [hunk 2]
```diff
@@ -26,4 +32,7 @@ in runCommand "llvm-binutils-${version}" { preferLocalBuild = true; } ''
    ln -s ${llvm}/bin/llvm-strip $out/bin/${prefix}strip
 
    ln -s ${lld}/bin/lld $out/bin/${prefix}ld
-''
+'' + lib.optionalString stdenv.targetPlatform.isWindows ''
+   ln -s ${llvm}/bin/llvm-windres $out/bin/${prefix}windres
+   ln -s ${llvm}/bin/llvm-dlltool $out/bin/${prefix}dlltool
+'')
```
Comment (sternenseemann) [inline pkgs/development/compilers/llvm/13/bintools/default.nix:38]: Is this relevant to other LLVM versions as well? Note that unless you add any changes to llvmPackages 14 and llvmPackages_git, it'll likely get lost in the future.

File: pkgs/development/libraries/libffi/default.nix [hunk 3]
```diff
@@ -45,6 +47,8 @@ stdenv.mkDerivation rec {
 
   checkInputs = [ dejagnu ];
 
+  nativeBuildInputs = lib.optional stdenv.hostPlatform.isWindows autoreconfHook;
```
Comment (sternenseemann) [inline pkgs/development/libraries/libffi/default.nix:50]: Can you add a comment why this is needed? I.e. is this a structural problem or will this fix itself in a future libffi release?

File: pkgs/development/libraries/gmp/6.x.nix [hunk 4]
```diff
@@ -29,7 +29,7 @@ let self = stdenv.mkDerivation rec {
   passthru.static = self.out;
 
   depsBuildBuild = [ buildPackages.stdenv.cc ];
-  nativeBuildInputs = [ m4 ];
+  nativeBuildInputs = [ m4 ] ++ lib.optional stdenv.hostPlatform.isWindows autoreconfHook;
```
Comment (sternenseemann) [inline pkgs/development/libraries/gmp/6.x.nix:32]: Can you add a comment why this is needed? I.e. is this a structural problem or will this fix itself in a future gmp release?

File: pkgs/development/compilers/llvm/13/libcxx/default.nix [hunk 5]
```diff
@@ -41,7 +41,13 @@ stdenv.mkDerivation rec {
...
       "-DLIBCXX_ENABLE_FILESYSTEM=OFF"
       "-DLIBCXX_ENABLE_EXCEPTIONS=OFF"
-    ] ++ lib.optional (!enableShared) "-DLIBCXX_ENABLE_SHARED=OFF";
+    ]
+    ++ lib.optional (!enableShared) "-DLIBCXX_ENABLE_SHARED=OFF"
+    ++ lib.optionals (!headersOnly && libcxxabi.semi-static) [
+      "-DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=TRUE"
+      "-DLIBCXX_CXX_ABI_LIBRARY_PATH=${libcxxabi}/lib"
+    ] ++ lib.optional (!headersOnly && libcxxabi.useLLVMUnwinder)
+      "-DLIBCXXABI_USE_LLVM_UNWINDER=ON";
```
Comment (sternenseemann) [inline pkgs/development/compilers/llvm/13/libcxx/default.nix:50]: Still confused about this one…

File: pkgs/development/compilers/llvm/13/libcxxabi/default.nix [hunk 6]
```diff
@@ -2,9 +2,14 @@
 , enableShared ? !stdenv.hostPlatform.isStatic
 , standalone ? stdenv.hostPlatform.useLLVM or false
 , withLibunwind ? !stdenv.isDarwin && !stdenv.isFreeBSD && !stdenv.hostPlatform.isWasm
-}:
+}: let
+  # lld doesn't support unresolved references on Windows https://github.com/llvm/llvm-project/issues/55245
+  semi-static = enableShared && stdenv.hostPlatform.isWindows && stdenv.cc.bintools.isLld;
 
-stdenv.mkDerivation rec {
+  enableShared' = enableShared && !semi-static;
```
Comment (sternenseemann) [inline pkgs/development/compilers/llvm/13/libcxxabi/default.nix:9]: Maybe onlyEnableShared is clearer?

**Review by sternenseemann — COMMENTED** [review]
Overall looks okay to me, but I can't really comment on the libtool patches and haven't tested it.

Doing things for LLVM >= 13 as long as it includes git is okay, may be nice to add it for 11 and 12 as well so it hits our default versions may be nice as well, but not a requirement.

**Comment by aakropotkin** [comment]
I reviewed those patches recently for `libtool` and rejected them because
they improperly handle flag specs, and will cause static linkage to become
high risk for ABI conflicts.

If you can be certain that those patches will only be applied to individual
derivations I think it lowers the risk, but be on the lookout for ABI
issues especially in C++.

**Comment by shlevy** [comment]
Here's the thread with the patches upstream that @aakropotkin mentions https://debbugs.gnu.org/cgi/bugreport.cgi?bug=27866

**Comment by sternenseemann** [comment]
Well, we definitely can't apply libtool patches that have been rejected upstream and cause incorrect behavior in some cases. If they are indeed required for mingw64-llvm cross, we should make them conditional to that platform. Unfortunately this means someone needs to keep an eye on cross-trunk whenever libtool is updated.

**Comment by aakropotkin** [comment]
This is likely the safest move.

I do think that these patches are useful to add as a setup hook or
alternate build for libtool that can be applied on a per-project basis.

The email thread has more details but the largest issue I saw was that
really this is a patch that fundamentally alters parts of libtool,
conflicting with the existing UI, for all platforms and tools in order to
resolve an issue in a specific Windows build env that is able to use two
compilers and two link editors.

If you fall into this specific category of user, and you understand the way
that the UI has been altered, then I think it's a very useful set of
patches; but as written they do not limit the scope of changes to the
relevant environment, they are not optionally enabled, and they do not
update the relevant documentation about the UI differences.

**Comment by Ericson2314** [comment]
Having interacted with some of the LLVM people there and you all, I would be happy to convene some meeting and hash out how things *ought* to work, if this a case of everyone hacking around each other.

**Comment by aakropotkin** [comment]
I'm up for it. A call or something would be good rather than an email chain
to avoid miscommunications and speed things along.

I'll be the first to admit that I really don't like how `libtool` handles
the flagspecs currently, and would really benefit from some help on
resolving it.

The biggest issue is honestly that I'm one person, and `libtool` patches
honestly just gets tested on my collection of VMs. I have Hydra building
which was a huge help, but no VMs there. Having more environments to run on
would be a life saver.

I have an idea of how I'd like libtool to handle this issue, but getting
there is daunting with the infrastructure I have set up currently.

If anyone could help make new tests as Hydra flake jobs that would even be
an enormous help. The existing test suite concerns me quite a bit, tons of
"remove this in 2009" with no further explanation floating around the
comments, and after driving it for the last six months I'm not convinced
that it's trustworthy; a lot of them seem to operate under the assumption
of FHS which is not ideal for today's containerized approaches. Having
Hydra jobs that just use libtool as a dependency to build something would
be significantly more useful.

**Comment by shlevy** [comment]
Draft until I can conditionalize libtool patches and respond to other comments.

**Comment by Ericson2314** [comment]
Did this get done in the meantime or something?

**Comment by shlevy** [comment]
No, too much complexity to finish quickly and the work project got cancelled.

**Pull request closed without merging**